ring/spread_minimizing_token_generator.go

```diff
@@ -0,0 +1,279 @@
+		token := allTokens[i]
+		if used[token] {
+			continue
+		}
+		tokens = append(tokens, token)
+	}
+	return tokens
+}
+
+// generateAllTokens generates the configured number of tokens (t.cfg.tokensPerInstance) for the underlying instance.
+// Generated tokens are sorted. Calls to this method will always return the same set of tokens.
+func (t *SpreadMinimizingTokenGenerator) generateAllTokens() Tokens {
+	firstInstanceTokens := t.generateFirstInstanceTokens()
+
+	if t.instanceID == 0 {
+		return firstInstanceTokens
+	}
+
+	tokensCount := t.cfg.TokensPerInstance
+	// tokensQueues is a slice of priority queues. Slice indexes correspond
+	// to the ids of instances, while priority queues represent the tokens
+	// of the corresponding instance, ordered from highest to lowest ownership.
+	tokensQueues := make([]ownershipPriorityQueue[ringToken], t.instanceID)
+
+	// Create and initialize priority queue of tokens for the first instance
+	tokensQueue := newPriorityQueue[ringToken](tokensCount)
+	prev := len(firstInstanceTokens) - 1
+	firstInstanceOwnership := 0.0
+	for tk, token := range firstInstanceTokens {
+		tokenOwnership := float64(getTokenDistance(firstInstanceTokens[prev], token))
+		firstInstanceOwnership += tokenOwnership
+		tokensQueue.Add(newRingTokenOwnershipInfo(token, firstInstanceTokens[prev]))
+		prev = tk
+	}
+	heap.Init(&tokensQueue)
+	tokensQueues[0] = tokensQueue
+
+	// instanceQueue is a priority queue of instances such that instances with higher ownership have a higher priority
+	instanceQueue := newPriorityQueue[ringInstance](t.instanceID)
+	instanceQueue.Add(newRingInstanceOwnershipInfo(0, firstInstanceOwnership))
+	heap.Init(&instanceQueue)
+
+	for i := 1; i <= t.instanceID; i++ {
+		optimalInstanceOwnership := float64(totalTokensCount) / float64(i+1)
+		currInstanceOwnership := 0.0
+		addedTokens := 0
+		// ignoredInstances is a slice of the current instances whose tokens
+		// don't have enough space to accommodate new tokens.
+		ignoredInstances := make([]ownershipInfo[ringInstance], 0, i)
+		tokens := make(Tokens, 0, tokensCount)
+		// currInstanceTokenQueue is the priority queue of tokens of newInstance
+		currInstanceTokenQueue := newPriorityQueue[ringToken](tokensCount)
+		for addedTokens < tokensCount {
+			optimalTokenOwnership := t.getOptimalTokenOwnership(optimalInstanceOwnership, currInstanceOwnership, uint32(tokensCount-addedTokens))
+			highestOwnershipInstance := instanceQueue.Peek()
+			if highestOwnershipInstance == nil || highestOwnershipInstance.ownership <= float64(optimalTokenOwnership) {
+				level.Warn(t.logger).Log("msg", "it was impossible to add a token because the instance with the highest ownership cannot satisfy the request", "added tokens", addedTokens+1, "highest ownership", highestOwnershipInstance.ownership, "requested ownership", optimalTokenOwnership)
+				// if this happens, we return the current set of tokens, although it is less than t.cfg.tokensPerInstance
+				break
```

Comment:
Now we need to panic here. We do have a test that checks that this doesn't happen up to an established number of instances, right? If so, let's just check that number of instances when we create this strategy, and panic here as it's unexpected.

Same applies to the `break` below.